### PR TITLE
Fix eaten async for function with labelled args

### LINKED ISSUE #674: async gets formatted away for functions with labeled arguments
```rescript
let f = async (a, b) => a + b

let f = async (~a, ~b) => a + b
```
gets reformatted to 
```rescript
let f = async (a, b) => a + b

let f = (~a, ~b) => a + b
```

### PR DESCRIPTION
**Pull request by mattdamon108**
This PR fixes #674

### CODE REVIEW

**Comment by mattdamon108** [comment]
I don't know the history of why letting the pexp_fun with nolabel arg keeps the attributes. https://github.com/rescript-lang/syntax/blob/a3c27ef3011e1c37ee3c2fbf8229f4dca49d3de7/src/res_parsetree_viewer.ml#L176

**Comment by mattdamon108** [comment]
This PR allows the Pexp_fun with labeled and optional to keep the `res.async` attribute only.

**Review by cknitt — APPROVED** [review]
Looks good to me, thanks a lot for fixing this! 🎉

(I don't know why Pexp_fun with Nolabel keeps all attributes either though.)

**Comment by cristianoc** [comment]
Oops, I had another fix done on the plane:
https://github.com/rescript-lang/syntax/pull/677

**Comment by cristianoc** [comment]
Basically, for `(@foo ~a) => ...` when one wants to put an attribute on the label, there's no way to represent this directly in the AST, as labelled args don't have attributes.
So the convention is to take the attributes on the function, and consider them applied to the label.

This does not fit well with the fact that `async` is put on the function, and some adjustment is needed.

File: src/res_parsetree_viewer.ml [hunk 1]
```diff
@@ -177,6 +184,17 @@ let funExpr expr =
...
+      pexp_desc = Pexp_fun (_, _defaultExpr, _pattern, _returnExpr);
+      pexp_attributes = attrs;
+    } as expr
+    when hasAsyncAttribute attrs ->
+    let asyncAttr =
+      List.filter
+        (fun (({txt}, _) : Parsetree.attribute) -> txt = "res.async")
+        attrs
+    in
+    collect asyncAttr [] {expr with pexp_attributes = []}
```
Comment (cristianoc) [inline src/res_parsetree_viewer.ml:197]: This throws away all the other attributes, if there are some.

**Comment by mattdamon108** [comment]
> 

There was a reason! Thank you for the explanation. I couldn't guess it. 👍

File: src/res_parsetree_viewer.ml (hunk 1 above)
Comment (mattdamon108) [reply src/res_parsetree_viewer.ml:197]: You are correct.

**Comment by mattdamon108** [comment]
I'm going to close this PR, #677 will fix the issue.

**Pull request closed without merging**